polar-core/src/data_filtering.rs

```diff
@@ -1158,3 +1223,447 @@ mod test {
+                            source = Self::joins(&self.types, source, rps.clone(), vec![])?;
+                        }
+
+                        Ok(source)
+                    }
+
+                    (PreDatum::Field(Proj(lss, _), lds), _)
+                    | (_, PreDatum::Field(Proj(lss, _), lds)) => {
+                        let mut source = source?;
+                        source =
+                            Self::joins(&self.types, source, lds.base.clone(), lds.path.clone())?;
+                        if let Source(lps) = &**lss {
+                            source = Self::joins(&self.types, source, lps.clone(), vec![])?;
+                        }
+                        Ok(source)
+                    }
+                    _ => source,
+                }
+            },
+        )?;
+        let rc = self
+            .constraints
+            .into_iter()
+            .fold(Ok(source), |source, con| match con {
+                (PreDatum::Field(lhs, _), kind, rhs) | (rhs, kind, PreDatum::Field(lhs, _)) => {
+                    let source = source?;
+                    let rhs = rhs.into();
+                    Ok(Rc::new(Select {
+                        kind,
+                        lhs,
+                        rhs,
+                        source,
+                    }))
+                }
+                _ => err_invalid("QueryInfo::into_filter()".to_string()),
+            })?;
+        Ok((*rc).clone())
+    }
+
+    fn path_disasm(t: &Term) -> PolarResult<Vec<String>> {
+        use Operator::*;
+        match t.value() {
+            Value::Expression(Operation {
+                operator: Dot,
+                args,
+            }) if args.len() == 2 && args[1].value().as_string().is_ok() => {
+                let mut p = Self::path_disasm(&args[0])?;
+                p.push(args[1].value().as_string().unwrap().to_string());
+                Ok(p)
+            }
+            Value::Variable(Symbol(s)) => Ok(vec![s.clone()]),
+            Value::ExternalInstance(ExternalInstance {
+                repr: Some(repr), ..
+            }) => {
+                eprintln!("{}", repr);
+                // FIXME hack for Ruby, obviously wrong
+                let re = regex::Regex::new(r"^#<([^:]+):.*").unwrap();
+                match re
+                    .captures(repr)
+                    .and_then(|cap| cap.get(1))
+                    .map(|c| c.as_str().to_owned())
+                {
+                    Some(tag) => Ok(vec![tag]),
+                    None => err_unimplemented(format!("QueryInfo::path_disasm({})", t.to_polar())),
+                }
+            }
+            _ => err_unimplemented(format!("QueryInfo::path_disasm({})", t.to_polar())),
+        }
+    }
+
+    fn data_source(&self, mut path: Vec<String>) -> DataSource {
+        let base = path.remove(0);
+        let base = self.entities.get(&vec![base]).unwrap().clone();
+        DataSource { base, path }
+    }
+
```

Comment:
It would be helpful to have some more comments on function definitions & bodies to make it easier to read, I added some locally, would you like me to push them up

Reply:
Sure!

Reply:
Bad merge conflicts now... I can help add / suggest once this is ready for another review.